src/core/QCheck.ml

```diff
@@ -763,16 +779,19 @@ module Shrink = struct
     | None -> ()
     | Some shrink -> list_elems shrink l yield
 
-  let string s yield =
+  let _shrink_string_and_bytes fold get s yield =
     let buf = Buffer.create 42 in
     list ~shrink:char
-      (string_fold_right (fun c acc -> c::acc) s [])
+      (fold (fun c acc -> c::acc) s [])
       (fun cs ->
          List.iter (fun c -> Buffer.add_char buf c) cs;
-         let s = Buffer.contents buf in
+         let s = get buf in
          Buffer.clear buf;
          yield s)
 
+  let bytes = _shrink_string_and_bytes bytes_fold_right Buffer.to_bytes
+  let string = _shrink_string_and_bytes string_fold_right Buffer.contents
```

Comment:
This touches the `string` shrinker and hence triggers an alarm bell in me :sweat_smile: 

The basic idea of the current `string` shrinker is
- to convert a `string` (or `bytes`) to a `list`
- fall back on the `list` shrinker
- map each shrunk `list` result back to a `string` (or `bytes`) using a `Buffer`.
    
1. Abstracting over the involved fold/get operations is certainly a valid approach to avoid code duplication. However I'm wondering whether it makes the shrinkers needlessly complex/higher-order :thinking:  
   Here's another way to obtain something similar without copy-pasting, which I think is clearer:
   Keeping the current `string` definition, define `bytes` as 
   ```ocaml
   let bytes b = Iter.map Bytes.of_string (string (Bytes.to_string b))
   ```
   By not touching the `string` shrinker we are sure not to affect its performance. Since `string` is more common than `bytes` this also optimizes for the common case.
   Alternatively, one could also keep an equivalent `bytes` version and define `string` in terms of it - perhaps using `Bytes.unsafe_to_string` to save copying during conversion (would happen for each shrinking candidate). 

2. I spent quite some time performance tweaking the shrinkers in #242 and #235 - even for the `string` shrinker (see https://github.com/c-cube/qcheck/pull/242#discussion_r867250149). I would thus be sad to see some of that hard-earned performance lost because, e.g., the compiler was unable to inline some of these function arguments. If we end up going with the `Shrink.string` changes I would therefore be grateful if you confirm with benchmarks that my concerns are unwarranted. Here are instructions for the current shrinker benchmarks:
   ```
   dune build test/core/shrink_benchmark.exe
   _build/default/test/core/shrink_benchmark.exe
   # shrinker benchmark output is now available in shrink_bench.log
   ```